File: android/src/main/kotlin/io/github/aswinsubhash/native_feed_player/NativeFeedPlayerPlugin.kt
package io.github.aswinsubhash.native_feed_player

import android.app.Activity
import android.app.Application
import android.content.ComponentCallbacks2
import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import java.util.concurrent.atomic.AtomicInteger

class NativeFeedPlayerPlugin : FlutterPlugin, ComponentCallbacks2, NativeFeedPlayerHostApi {
    private companion object {
        private const val VIDEO_VIEW_TYPE = "native_feed_player/video_view"

        /** Process-wide IDs prevent collisions after engine reattachment. */
        private val controllerIdSeed = AtomicInteger(0)
    }

    private var appContext: Context? = null
    private var exoPlayerManager: ExoPlayerManager? = null
    private var binaryMessenger: BinaryMessenger? = null

    private val stateEvents = BufferedStreamHandler<PlaybackStateEvent>()
    private val positionEvents = BufferedStreamHandler<PositionEvent>()
    private val metricsEvents = BufferedStreamHandler<MetricsEvent>()
    private val videoSizeEvents = BufferedStreamHandler<VideoSizeEvent>()
    private val lifecycleEvents = BufferedStreamHandler<ControllerLifecycleEvent>()

    /** Started-activity count used for foreground playback state. */
    private var startedActivityCount = 0
    private var activityCallbacks: Application.ActivityLifecycleCallbacks? = null

    private val textureViewPool = TextureViewPool(maxPoolSize = 8)
    private val videoViews = PlatformViewRegistry<Int, NativeVideoPlatformView>()
    private val attachedControllerByViewId = mutableMapOf<Int, Int>()
    private var textureOutputs: TextureOutputRegistry? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        appContext = flutterPluginBinding.applicationContext
        appContext?.registerComponentCallbacks(this)
        binaryMessenger = flutterPluginBinding.binaryMessenger
        textureOutputs = TextureOutputRegistry(flutterPluginBinding.textureRegistry)

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            VIDEO_VIEW_TYPE,
            NativeVideoViewFactory(
                textureViewPool = textureViewPool,
                onCreate = { viewId, view -> handleVideoViewCreated(viewId, view) },
                onDispose = { viewId, view -> handleVideoViewDisposed(viewId, view) }
            )
        )

        PlaybackStateEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            PlaybackStateStreamAdapter(stateEvents)
        )
        PositionEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            PositionStreamAdapter(positionEvents)
        )
        MetricsEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            MetricsStreamAdapter(metricsEvents)
        )
        VideoSizeEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            VideoSizeStreamAdapter(videoSizeEvents)
        )
        LifecycleEventsStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            LifecycleStreamAdapter(lifecycleEvents)
        )

        exoPlayerManager = ExoPlayerManager(
            context = flutterPluginBinding.applicationContext,
            onState = { controllerId, status, error ->
                stateEvents.emit(
                    PlaybackStateEvent(
                        controllerId = controllerId.toLong(),
                        status = status,
                        error = error
                    )
                )
            },
            onReleased = { controllerId, reason ->
                lifecycleEvents.emit(
                    ControllerLifecycleEvent(
                        controllerId = controllerId.toLong(),
                        reason = reason
                    )
                )
            },
            onPosition = { event -> positionEvents.emit(event) },
            onMetrics = { event -> metricsEvents.emit(event) },
            onVideoSize = { event -> videoSizeEvents.emit(event) }
        )

        registerActivityCallbacks(flutterPluginBinding.applicationContext)

        NativeFeedPlayerHostApi.setUp(
            binaryMessenger = flutterPluginBinding.binaryMessenger,
            api = this
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        unregisterActivityCallbacks()
        releaseTextureOutputs()
        textureOutputs = null
        appContext?.unregisterComponentCallbacks(this)
        appContext = null
        exoPlayerManager?.disposeAll()
        exoPlayerManager = null
        videoViews.clear()
        attachedControllerByViewId.clear()
        textureViewPool.clear()
        stateEvents.detach()
        positionEvents.detach()
        metricsEvents.detach()
        lifecycleEvents.detach()
        NativeFeedPlayerHostApi.setUp(binaryMessenger = binding.binaryMessenger, api = null)
        binaryMessenger = null
    }

    override fun initialize(config: FeedPlayerConfigMessage) {
        releaseTextureOutputs()
        attachedControllerByViewId.clear()
        managerOrThrow().initialize(config)
        stateEvents.clearPending()
        positionEvents.clearPending()
        metricsEvents.clearPending()
        videoSizeEvents.clearPending()
        lifecycleEvents.clearPending()
    }

    override fun setSources(sources: List<FeedSourceMessage>) {
        managerOrThrow().setSources(sources.map { it.toRegisteredSource() })
    }

    override fun appendSources(sources: List<FeedSourceMessage>) {
        managerOrThrow().appendSources(sources.map { it.toRegisteredSource() })
    }

    override fun removeSources(request: SourceIdsRequest) {
        managerOrThrow().removeSources(request.sourceIds)
    }

    override fun createController(request: CreateControllerRequest): Long {
        if (request.sourceId.isBlank()) {
            throw FlutterError("invalid_source", "createController requires a source id.", null)
        }
        val controllerId = controllerIdSeed.incrementAndGet()
        managerOrThrow().createController(
            controllerId = controllerId,
            sourceId = request.sourceId,
            autoPlay = request.autoPlay,
            looping = request.looping
        )
        return controllerId.toLong()
    }

    override fun disposeController(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().disposeController(controllerId)
        }
    }

    override fun play(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().play(controllerId)
        }
    }

    override fun pause(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().pause(controllerId)
        }
    }

    override fun seekTo(request: SeekRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId > 0) {
            managerOrThrow().seekTo(controllerId, request.positionMs)
        }
    }

    override fun setVolume(request: ControllerDoubleRequest) {
        managerOrThrow().setVolume(request.controllerId.toInt(), request.value)
    }

    override fun setMuted(request: ControllerFlagRequest) {
        managerOrThrow().setMuted(request.controllerId.toInt(), request.value)
    }

    override fun setPlaybackSpeed(request: ControllerDoubleRequest) {
        managerOrThrow().setPlaybackSpeed(request.controllerId.toInt(), request.value)
    }

    override fun setLooping(request: ControllerFlagRequest) {
        managerOrThrow().setLooping(request.controllerId.toInt(), request.value)
    }

    override fun setAudioPolicy(policy: AudioPolicyMessage) {
        managerOrThrow().applyAudioPolicy(policy)
    }

    override fun setVisibleSource(request: VisibleSourceRequest) {
        managerOrThrow().setVisibleSource(request.sourceId)
    }

    override fun evictCachedMedia(request: SourceIdsRequest) {
        managerOrThrow().evictCachedMedia(request.sourceIds)
    }

    override fun clearMediaCache() {
        managerOrThrow().clearMediaCache()
    }

    override fun cacheStatus(request: VisibleSourceRequest): CacheStatusMessage =
        managerOrThrow().cacheStatus(request.sourceId)

    override fun cacheUsageBytes(): Long = managerOrThrow().cacheUsageBytes()

    override fun attachView(request: AttachViewRequest) {
        val controllerId = request.controllerId.toInt()
        val viewId = request.viewId.toInt()
        if (controllerId <= 0 || viewId < 0) {
            throw FlutterError(
                "invalid_attach",
                "attachView requires valid controllerId and viewId.",
                null
            )
        }

        val view = videoViews[viewId]
            ?: throw FlutterError("view_not_found", "No video view found for id=$viewId.", null)

        val manager = managerOrThrow()
        val previousController = attachedControllerByViewId[viewId]
        if (previousController != null && previousController != controllerId) {
            manager.detachControllerFromView(previousController)
        }
        attachedControllerByViewId[viewId] = controllerId
        manager.attachControllerToView(controllerId, view.textureView)
    }

    override fun detachView(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        if (controllerId <= 0) {
            return
        }
        val manager = managerOrThrow()
        manager.detachControllerFromView(controllerId)
        val staleViewIds = attachedControllerByViewId
            .filterValues { it == controllerId }
            .keys
            .toList()
        for (viewId in staleViewIds) {
            attachedControllerByViewId.remove(viewId)
        }
    }

    override fun attachTexture(request: ControllerRequest): Long {
        val controllerId = request.controllerId.toInt()
        val registry = textureOutputs
            ?: throw FlutterError("not_attached", "No texture registry available.", null)
        val manager = managerOrThrow()
        return registry.attach(controllerId) { surface ->
            manager.bindSurface(controllerId, surface)
        }
    }

    override fun detachTexture(request: ControllerRequest) {
        val controllerId = request.controllerId.toInt()
        val manager = exoPlayerManager
        textureOutputs?.detach(controllerId) { _ ->
            manager?.unbindSurface(controllerId)
        }
    }

    override fun disposeAll() {
        releaseTextureOutputs()
        managerOrThrow().disposeAll()
        attachedControllerByViewId.clear()
    }

    private fun releaseTextureOutputs() {
        val manager = exoPlayerManager
        textureOutputs?.clear { controllerId, _ -> manager?.unbindSurface(controllerId) }
    }

    override fun onTrimMemory(level: Int) {
        exoPlayerManager?.onTrimMemory(level)
    }

    @Deprecated("ComponentCallbacks2.onLowMemory is deprecated by the platform")
    override fun onLowMemory() {
        exoPlayerManager?.onLowMemory()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        // No-op: handled by the Flutter engine and ExoPlayer internally.
    }

    private fun registerActivityCallbacks(context: Context) {
        val application = context.applicationContext as? Application ?: return
        unregisterActivityCallbacks()

        val callbacks = object : Application.ActivityLifecycleCallbacks {
            override fun onActivityStarted(activity: Activity) {
                startedActivityCount += 1
                if (startedActivityCount == 1) {
                    exoPlayerManager?.onAppForegrounded()
                }
            }

            override fun onActivityStopped(activity: Activity) {
                startedActivityCount = (startedActivityCount - 1).coerceAtLeast(0)
                if (startedActivityCount == 0) {
                    exoPlayerManager?.onAppBackgrounded()
                }
            }

            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) = Unit
            override fun onActivityResumed(activity: Activity) = Unit
            override fun onActivityPaused(activity: Activity) = Unit
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) = Unit
            override fun onActivityDestroyed(activity: Activity) = Unit
        }
        application.registerActivityLifecycleCallbacks(callbacks)
        activityCallbacks = callbacks
    }

    private fun unregisterActivityCallbacks() {
        val application = appContext?.applicationContext as? Application
        activityCallbacks?.let { application?.unregisterActivityLifecycleCallbacks(it) }
        activityCallbacks = null
        startedActivityCount = 0
    }

    private fun managerOrThrow(): ExoPlayerManager {
        return exoPlayerManager
            ?: throw FlutterError(
                "not_attached",
                "NativeFeedPlayerPlugin is not attached to a Flutter engine.",
                null
            )
    }

    private fun handleVideoViewCreated(viewId: Int, view: NativeVideoPlatformView) {
        val previousControllerId = attachedControllerByViewId.remove(viewId)
        if (previousControllerId != null) {
            exoPlayerManager?.detachControllerFromView(previousControllerId)
        }
        videoViews.register(viewId, view)
    }

    private fun handleVideoViewDisposed(viewId: Int, view: NativeVideoPlatformView) {
        if (!videoViews.removeIfCurrent(viewId, view)) {
            textureViewPool.release(view.textureView)
            return
        }
        val controllerId = attachedControllerByViewId.remove(viewId)
        if (controllerId != null) {
            exoPlayerManager?.detachControllerFromView(controllerId)
        }
        textureViewPool.release(view.textureView)
    }
}

private fun FeedSourceMessage.toRegisteredSource(): RegisteredSource = RegisteredSource(
    id = id,
    uri = uri,
    rank = rank.toInt(),
    kind = kind,
    headers = headers
)

File: ios/native_feed_player/Sources/native_feed_player/NativeFeedPlayerPlugin.swift
import Flutter
import Foundation
import UIKit

/// Buffers events until the Dart event stream attaches.
final class BufferedEventSink<T> {
  private let maxBuffered: Int
  private var pending: [T] = []
  private var sink: PigeonEventSink<T>?

  init(maxBuffered: Int = 64) {
    self.maxBuffered = maxBuffered
  }

  func attach(_ sink: PigeonEventSink<T>) {
    self.sink = sink
    let buffered = pending
    pending.removeAll()
    for event in buffered {
      sink.success(event)
    }
  }

  func detach() {
    sink = nil
    pending.removeAll()
  }

  func clearPending() {
    pending.removeAll()
  }

  func emit(_ event: T) {
    if let sink {
      sink.success(event)
      return
    }
    if pending.count >= maxBuffered {
      pending.removeFirst()
    }
    pending.append(event)
  }
}

/// Adapts a Pigeon event channel to `BufferedEventSink`.
private final class PlaybackStateStreamAdapter: PlaybackStateEventsStreamHandler {
  private let holder: BufferedEventSink<PlaybackStateEvent>

  init(holder: BufferedEventSink<PlaybackStateEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<PlaybackStateEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class PositionStreamAdapter: PositionEventsStreamHandler {
  private let holder: BufferedEventSink<PositionEvent>

  init(holder: BufferedEventSink<PositionEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<PositionEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class MetricsStreamAdapter: MetricsEventsStreamHandler {
  private let holder: BufferedEventSink<MetricsEvent>

  init(holder: BufferedEventSink<MetricsEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<MetricsEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class VideoSizeStreamAdapter: VideoSizeEventsStreamHandler {
  private let holder: BufferedEventSink<VideoSizeEvent>

  init(holder: BufferedEventSink<VideoSizeEvent>) {
    self.holder = holder
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<VideoSizeEvent>) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

private final class LifecycleStreamAdapter: LifecycleEventsStreamHandler {
  private let holder: BufferedEventSink<ControllerLifecycleEvent>

  init(holder: BufferedEventSink<ControllerLifecycleEvent>) {
    self.holder = holder
  }

  override func onListen(
    withArguments arguments: Any?,
    sink: PigeonEventSink<ControllerLifecycleEvent>
  ) {
    holder.attach(sink)
  }

  override func onCancel(withArguments arguments: Any?) {
    holder.detach()
  }
}

public final class NativeFeedPlayerPlugin: NSObject, FlutterPlugin, NativeFeedPlayerHostApi {
  private static let videoViewType = "native_feed_player/video_view"

  /// Process-wide ID seed retained across engine reattachment.
  private static var controllerIdSeed = 0

  private let stateEvents = BufferedEventSink<PlaybackStateEvent>()
  private let positionEvents = BufferedEventSink<PositionEvent>()
  private let metricsEvents = BufferedEventSink<MetricsEvent>()
  private let videoSizeEvents = BufferedEventSink<VideoSizeEvent>()
  private let lifecycleEvents = BufferedEventSink<ControllerLifecycleEvent>()

  private var manager: AVPlayerManager?
  private var memoryWarningObserver: NSObjectProtocol?
  private let renderViewPool = RenderViewPool(maxPoolSize: 8)
  private let videoViews = PlatformViewRegistry<Int64, NativeVideoPlatformView>()
  private var attachedControllerByViewId: [Int64: Int] = [:]
  private var binaryMessenger: FlutterBinaryMessenger?
  private var textureOutputs: TextureOutputRegistry?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = NativeFeedPlayerPlugin()
    instance.binaryMessenger = registrar.messenger()
    instance.textureOutputs = TextureOutputRegistry(registry: registrar.textures())
    instance.configure(messenger: registrar.messenger())

    let viewFactory = NativeVideoViewFactory(
      renderViewPool: instance.renderViewPool,
      onCreate: { [weak instance] viewId, view in
        instance?.handleVideoViewCreated(viewId: viewId, view: view)
      },
      onDispose: { [weak instance] viewId, view in
        instance?.handleVideoViewDisposed(viewId: viewId, view: view)
      }
    )
    registrar.register(viewFactory, withId: videoViewType)
    NativeFeedPlayerHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
  }

  deinit {
    if let observer = memoryWarningObserver {
      NotificationCenter.default.removeObserver(observer)
      memoryWarningObserver = nil
    }
    if let binaryMessenger {
      NativeFeedPlayerHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: nil)
    }
    textureOutputs?.clear()
    manager?.disposeAll()
    videoViews.removeAll()
    attachedControllerByViewId.removeAll()
    renderViewPool.clear()
  }

  // MARK: - Host API

  func initialize(config: FeedPlayerConfigMessage) throws {
    textureOutputs?.clear()
    attachedControllerByViewId.removeAll()
    try managerOrThrow().initialize(config: config)
    stateEvents.clearPending()
    positionEvents.clearPending()
    metricsEvents.clearPending()
    videoSizeEvents.clearPending()
    lifecycleEvents.clearPending()
  }

  func setSources(sources: [FeedSourceMessage]) throws {
    do {
      try managerOrThrow().setSources(sources.map { $0.toRegisteredSource() })
    } catch let error as AVPlayerManager.PlaybackSetupError {
      throw PigeonError(code: error.code, message: error.message, details: nil)
    }
  }

  func appendSources(sources: [FeedSourceMessage]) throws {
    do {
      try managerOrThrow().appendSources(sources.map { $0.toRegisteredSource() })
    } catch let error as AVPlayerManager.PlaybackSetupError {
      throw PigeonError(code: error.code, message: error.message, details: nil)
    }
  }

  func removeSources(request: SourceIdsRequest) throws {
    try managerOrThrow().removeSources(request.sourceIds)
  }

  func createController(request: CreateControllerRequest) throws -> Int64 {
    if request.sourceId.isEmpty {
      throw PigeonError(
        code: "invalid_source",
        message: "createController requires a source id.",
        details: nil
      )
    }

    NativeFeedPlayerPlugin.controllerIdSeed += 1
    let controllerId = NativeFeedPlayerPlugin.controllerIdSeed

    do {
      try managerOrThrow().createController(
        controllerId: controllerId,
        sourceId: request.sourceId,
        autoPlay: request.autoPlay,
        looping: request.looping
      )
    } catch let setupError as AVPlayerManager.PlaybackSetupError {
      throw PigeonError(code: setupError.code, message: setupError.message, details: nil)
    } catch {
      throw PigeonError(
        code: "create_failed",
        message: error.localizedDescription,
        details: nil
      )
    }

    return Int64(controllerId)
  }

  func disposeController(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().disposeController(controllerId: controllerId)
    }
  }

  func play(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().play(controllerId: controllerId)
    }
  }

  func pause(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().pause(controllerId: controllerId)
    }
  }

  func seekTo(request: SeekRequest) throws {
    let controllerId = Int(request.controllerId)
    if controllerId > 0 {
      try managerOrThrow().seekTo(controllerId: controllerId, positionMs: request.positionMs)
    }
  }

  func setVolume(request: ControllerDoubleRequest) throws {
    try managerOrThrow().setVolume(controllerId: Int(request.controllerId), value: request.value)
  }

  func setMuted(request: ControllerFlagRequest) throws {
    try managerOrThrow().setMuted(controllerId: Int(request.controllerId), value: request.value)
  }

  func setPlaybackSpeed(request: ControllerDoubleRequest) throws {
    try managerOrThrow().setPlaybackSpeed(
      controllerId: Int(request.controllerId),
      speed: request.value
    )
  }

  func setLooping(request: ControllerFlagRequest) throws {
    try managerOrThrow().setLooping(
      controllerId: Int(request.controllerId),
      looping: request.value
    )
  }

  func setAudioPolicy(policy: AudioPolicyMessage) throws {
    try managerOrThrow().applyAudioPolicy(policy)
  }

  func setVisibleSource(request: VisibleSourceRequest) throws {
    try managerOrThrow().setVisibleSource(request.sourceId)
  }

  func evictCachedMedia(
    request: SourceIdsRequest,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    do {
      try managerOrThrow().evictCachedMedia(request.sourceIds) {
        DispatchQueue.main.async { completion(.success(())) }
      }
    } catch {
      DispatchQueue.main.async { completion(.failure(error)) }
    }
  }

  func clearMediaCache(completion: @escaping (Result<Void, Error>) -> Void) {
    do {
      try managerOrThrow().clearMediaCache {
        DispatchQueue.main.async { completion(.success(())) }
      }
    } catch {
      DispatchQueue.main.async { completion(.failure(error)) }
    }
  }

  func cacheStatus(
    request: VisibleSourceRequest,
    completion: @escaping (Result<CacheStatusMessage, Error>) -> Void
  ) {
    do {
      try managerOrThrow().cacheStatus(sourceId: request.sourceId) { status in
        DispatchQueue.main.async { completion(.success(status)) }
      }
    } catch {
      DispatchQueue.main.async { completion(.failure(error)) }
    }
  }

  func cacheUsageBytes(completion: @escaping (Result<Int64, Error>) -> Void) {
    do {
      try managerOrThrow().cacheUsageBytes { bytes in
        DispatchQueue.main.async { completion(.success(bytes)) }
      }
    } catch {
      DispatchQueue.main.async { completion(.failure(error)) }
    }
  }

  func attachView(request: AttachViewRequest) throws {
    let controllerId = Int(request.controllerId)
    let viewId = Int64(request.viewId)
    guard controllerId > 0, viewId >= 0 else {
      throw PigeonError(
        code: "invalid_attach",
        message: "attachView requires valid controllerId and viewId.",
        details: nil
      )
    }
    guard let platformView = videoViews[viewId] else {
      throw PigeonError(
        code: "view_not_found",
        message: "No video view found for id=\(viewId).",
        details: nil
      )
    }

    let manager = try managerOrThrow()
    if let previous = attachedControllerByViewId[viewId], previous != controllerId {
      manager.detach(controllerId: previous)
    }
    // A controller can render into only one platform view. Remove stale plugin
    // mappings when Flutter reuses/replaces a view before disposal arrives.
    let staleViewIds = attachedControllerByViewId.compactMap { otherViewId, attached in
      attached == controllerId && otherViewId != viewId ? otherViewId : nil
    }
    for staleViewId in staleViewIds {
      videoViews[staleViewId]?.renderView.setPlayer(nil)
      attachedControllerByViewId.removeValue(forKey: staleViewId)
    }
    attachedControllerByViewId[viewId] = controllerId
    manager.attach(controllerId: controllerId, renderView: platformView.renderView)
  }

  func detachView(request: ControllerRequest) throws {
    let controllerId = Int(request.controllerId)
    guard controllerId > 0 else {
      return
    }
    let manager = try managerOrThrow()
    manager.detach(controllerId: controllerId)
    let attachedViewIds = attachedControllerByViewId.compactMap { viewId, attached in
      attached == controllerId ? viewId : nil
    }
    for viewId in attachedViewIds {
      videoViews[viewId]?.renderView.setPlayer(nil)
      attachedControllerByViewId.removeValue(forKey: viewId)
    }
  }

  func attachTexture(request: ControllerRequest) throws -> Int64 {
    let controllerId = Int(request.controllerId)
    guard let textureOutputs else {
      throw PigeonError(
        code: "not_attached",
        message: "No texture registry available.",
        details: nil
      )
    }
    let manager = try managerOrThrow()
    guard let player = manager.player(for: controllerId) else {
      throw PigeonError(
        code: "controller_not_found",
        message: "No live controller with id=\(controllerId).",
        details: nil
      )
    }
    return textureOutputs.attach(
      controllerId: controllerId,
      player: player,
      onFirstPixel: { [weak manager] in
        manager?.markTextureFirstFrame(controllerId: controllerId)
      }
    )
  }

  func detachTexture(request: ControllerRequest) throws {
    textureOutputs?.detach(controllerId: Int(request.controllerId))
  }

  func disposeAll() throws {
    textureOutputs?.clear()
    try managerOrThrow().disposeAll()
    attachedControllerByViewId.removeAll()
  }

  // MARK: - Wiring

  private func configure(messenger: FlutterBinaryMessenger) {
    PlaybackStateEventsStreamHandler.register(
      with: messenger,
      streamHandler: PlaybackStateStreamAdapter(holder: stateEvents)
    )
    PositionEventsStreamHandler.register(
      with: messenger,
      streamHandler: PositionStreamAdapter(holder: positionEvents)
    )
    MetricsEventsStreamHandler.register(
      with: messenger,
      streamHandler: MetricsStreamAdapter(holder: metricsEvents)
    )
    VideoSizeEventsStreamHandler.register(
      with: messenger,
      streamHandler: VideoSizeStreamAdapter(holder: videoSizeEvents)
    )
    LifecycleEventsStreamHandler.register(
      with: messenger,
      streamHandler: LifecycleStreamAdapter(holder: lifecycleEvents)
    )

    manager = AVPlayerManager(
      onState: { [weak self] controllerId, status, error in
        self?.onMain {
          self?.stateEvents.emit(
            PlaybackStateEvent(
              controllerId: Int64(controllerId),
              status: status,
              error: error
            )
          )
        }
      },
      onReleased: { [weak self] controllerId, reason in
        self?.onMain {
          self?.handleManagerRelease(controllerId: controllerId, reason: reason)
        }
      },
      onPosition: { [weak self] event in
        self?.onMain { self?.positionEvents.emit(event) }
      },
      onMetrics: { [weak self] event in
        self?.onMain { self?.metricsEvents.emit(event) }
      },
      onVideoSize: { [weak self] event in
        self?.onMain { self?.videoSizeEvents.emit(event) }
      }
    )

    if let observer = memoryWarningObserver {
      NotificationCenter.default.removeObserver(observer)
      memoryWarningObserver = nil
    }
    memoryWarningObserver = NotificationCenter.default.addObserver(
      forName: UIApplication.didReceiveMemoryWarningNotification,
      object: nil,
      queue: .main
    ) { [weak self] _ in
      self?.manager?.onMemoryWarning()
    }
  }

  private func onMain(_ work: @escaping () -> Void) {
    if Thread.isMainThread {
      work()
    } else {
      DispatchQueue.main.async(execute: work)
    }
  }

  /// A manager release must sever every Flutter rendering edge before Dart is
  /// told that the controller is gone.
  private func handleManagerRelease(controllerId: Int, reason: ReleaseReasonMessage) {
    textureOutputs?.detach(controllerId: controllerId)
    let staleViewIds = attachedControllerByViewId.compactMap { viewId, attached in
      attached == controllerId ? viewId : nil
    }
    for viewId in staleViewIds {
      videoViews[viewId]?.renderView.setPlayer(nil)
      attachedControllerByViewId.removeValue(forKey: viewId)
    }
    lifecycleEvents.emit(
      ControllerLifecycleEvent(controllerId: Int64(controllerId), reason: reason)
    )
  }

  private func handleVideoViewCreated(viewId: Int64, view: NativeVideoPlatformView) {
    if let previousControllerId = attachedControllerByViewId.removeValue(forKey: viewId) {
      manager?.detach(controllerId: previousControllerId)
    }
    videoViews.register(view, for: viewId)
  }

  private func handleVideoViewDisposed(viewId: Int64, view: NativeVideoPlatformView) {
    guard videoViews.removeIfCurrent(view, for: viewId) else {
      renderViewPool.release(view.renderView)
      return
    }
    if let controllerId = attachedControllerByViewId.removeValue(forKey: viewId) {
      manager?.detach(controllerId: controllerId, renderView: view.renderView)
    }
    renderViewPool.release(view.renderView)
  }

  private func managerOrThrow() throws -> AVPlayerManager {
    guard let manager else {
      throw PigeonError(
        code: "not_attached",
        message: "NativeFeedPlayerPlugin is not attached to a Flutter engine.",
        details: nil
      )
    }
    return manager
  }
}

extension FeedSourceMessage {
  fileprivate func toRegisteredSource() -> RegisteredSource {
    RegisteredSource(
      id: id,
      uri: uri,
      rank: Int(rank),
      kind: kind,
      headers: headers
    )
  }
}
